index.js

```diff
@@ -48,50 +48,274 @@ const player = {
+    }
+  }
+  searchResults.songs.sort(sortTitlesAlphabetically)
+  searchResults.playlists.sort(sortNameAlphabetically)
+  return searchResults
 }
 
 function searchByDuration(duration) {
-  // your code here
+  //function will get duration proximity by subtracting durations. the song/playlist with the smallest duration difference (closest to zero) will be chosen.
+  let durationInSeconds = convertMinutesToSeconds(duration)
+  // assign temporary value to the variable that will contain the closest match
+  let currentClosestDuration = Number.MAX_SAFE_INTEGER
+  let match = null
+  for (let song of player.songs) {
+    let durationsDifference = Math.abs(durationInSeconds - song.duration)
+    if (durationsDifference < currentClosestDuration) {
+      currentClosestDuration = durationsDifference
+      match = song
+    }
+  }
+  for (let playlist of player.playlists) {
+    let currentPlaylistDuration = playlistDuration(playlist.id)
+    let durationsDifference = Math.abs(
+      durationInSeconds - currentPlaylistDuration
+    )
+    if (durationsDifference < currentClosestDuration) {
+      currentClosestDuration = durationsDifference
+      match = playlist
+    }
+  }
+  return match
+}
+
+//UTILITY FUNCTIONS
+
+//convert from second to mm:ss format. e.g. 160 to 02:40
+function convertSecondsToMinutes(time) {
+  assertIsNumber(time)
+  let minutes = Math.floor(time / 60)
+  let seconds = time - minutes * 60
+  let paddedMinutes = minutes.toString().padStart(2, 0)
+  let paddedSeconds = seconds.toString().padStart(2, 0)
+  return `${paddedMinutes}:${paddedSeconds}`
+}
+
+//convert from mm:ss format to seconds. e.g. 02:40 to 160
+function convertMinutesToSeconds(time) {
+  //mmssRe matches mm:ss and allows more than two minute digits.
+  let mmssRe = new RegExp(/(^\d{2,})[:](\d{2}$)/)
+  let matches = time.match(mmssRe)
+  if (!matches) {
+    throw new Error(`Oy vey! time entered has to be in the mm:ss format! Time entered: ${time}`)
+  }
+
+  let seconds = parseInt(matches[2])
+  let minutes = parseInt(matches[1])
+  return seconds + minutes * 60
+}
+
+function getSongById(id) {
+  assertIsNumber(id)
+  for (let song of player.songs) {
+    if (song.id === id) {
+      return song
+    }
+  }
+  throw new Error(
+    `Whoops! we couldn't find a song that matches the ID you entered. Song ID entered: ${id}`
+    )
+  }
+  
+function getSongIndexById(id) {
+  let song = getSongById(id)
+  return player.songs.indexOf(song)
+}
+
+function getPlaylistById(id) {
+  assertIsNumber(id)
+  for (let playlist of player.playlists) {
+    if (playlist.id === id) {
+      return playlist
+    }
+  }
+  throw new Error(
+    `Hmmm.. There's no playlist with that ID. The playlist ID entered: ${id}`
+    )
+  }
+  
+//gets number (i) and goes through all the songs / playlists to see if anyone has it. if not, it is considered avaliable.
+function getVacantId(array) {
+  mainLoop: for (let i = 1; i <= array.length + 1; i++) {
```

Comment:
Woah! I literally just learned something new.
Never saw one of these labels before...